File: algorithm/go/1857-largestColorValueInADirectedGraph.go
// go: build ignore
package main

import "fmt"

func boolToInt(b bool) int {
	if !b {
		return 0
	}
	return 1
}

func largestPathValue(colors string, edges [][]int) int {
    n := len(colors)
	adj := make([][]int, n)
	rank := make([]int, n)

	for _, e := range edges {
		adj[e[0]] = append(adj[e[0]], e[1])
		rank[e[1]]++
	}

	dp := make([][]int, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]int, 26)
	}

	q := []int{}
	for u := range n {
		if rank[u] == 0 {
			q = append(q, u)
			dp[u][int(colors[u] - 'a')]++
		}
	}

	visited := 0
	res := 0

	for len(q) > 0 {
		u := q[0]
		q = q[1:]
		visited++

		for _, v := range adj[u] {
			for c := range 26 {
				tmp := boolToInt(int(colors[v] - 'a') - c == 0)
				dp[v][c] = max(dp[v][c], dp[u][c] + tmp)
			}
			
			rank[v]--
			if rank[v] == 0 {
				q = append(q, v)
			}
		}

		for _, val := range dp[u] {
			res = max(res, val)
		}
	}

	if visited != n {
		return -1
	}
	return res
}

func main() {
	colors := "abaca"
	edges := [][]int{{0, 1}, {0, 2}, {2, 3}, {3, 4}}
	fmt.Println(largestPathValue(colors, edges))
}

File: leetcode/cpp/1857-largestColorValueInADirectedGraph.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
	private:
		int dfs(int u, string& colors, vector<vector<int>>& adj, vector<vector<int>>& count, vector<int>& visited) {
			if (visited[u] == 0) {
				visited[u] = 1;
				for (int v: adj[u]) {
					if (dfs(v, colors, adj, count, visited) == INT_MAX) {
						return INT_MAX;
					}

					for (int c = 0; c < 26; c++) {
						count[u][c] = max(count[u][c], count[v][c]);
					}
				}

				++count[u][colors[u] - 'a'];
				visited[u] = 2;
			}

			return visited[u] == 2 ? count[u][colors[u] - 'a'] : INT_MAX;
		}

    public:
        int largestPathValue(string colors, vector<vector<int>>& edges) {
            int n = colors.length();

            vector<vector<int>> adj(n);
            vector<vector<int>> count(n, vector<int>(26, 0));
			vector<int> visited(n);

            for (auto& e: edges) {
                adj[e[0]].push_back(e[1]);
			}

			int ans = 0;
			for (int i = 0; ans != INT_MAX && i < n; i++) {
				ans = max(ans, dfs(i, colors, adj, count, visited));
			}

			return ans == INT_MAX ? -1 : ans;
		}
};

int main() {
    Solution sol;
    string colors = "abaca";
    vector<vector<int>> edges = {{0,1},{0,2},{2,3},{3,4}};
    
    cout << sol.largestPathValue(colors, edges) << endl; 
    return 0;
}
